### PR TITLE
Remove afterEvaluate

### PR DESCRIPTION
**Pull request by hfhbd**
This PR removes the afterEvaluate, mostly used with android and Kotlin multiplatform support by reconfiguring existing tasks using providers.

Advantage: 
- No afterEvaluate anymore makes the Kotlin DSL usage more friendly, because Gradle generates typesafe task accessors.

Disadvantages:
- Only with android and mpp: there are now multiple licensee tasks for each android variant: `licenseeDebug` and `licenseeAndroidDebug`. While the first one is disabled, it still exists and visible. AFAIK, there is no API to remove a Gradle task.
- Without afterEvaluate, there is no check possible to warn the user if licensee is applied without a language. Instead, it will fail: `task licensee not found` when applying without a language plugin.

What do you think?

### CODE REVIEW

**Comment by JakeWharton** [comment]
Well I'm not really thrilled with either choice. That should be Gradle's slogan.

Does this mean that if you disable a variant in an Android project you still see its tasks created? Is there a new AGP API that we can use to fix that case? I know there's the "before" and "on" callbacks now.

**Comment by hfhbd** [comment]
Yeah, the whole thing, AGP + KGP, is a mess! 
The biggest problem is Gradle: If you want to remove `afterEvaluate`, you can only use `withPlugin`, because `hasPlugin` depends on the execution order of the plugins, so it wouldn't run if android/kgp is applied later. But `withPlugin` is basically a fire-and-forget, (ideally) you can't track its execution or the absence of one plugin. So the only workaround is having multiple nested `withPlugin`s, resulting in the reconfiguration of existing tasks...
(You could hack it with AtomicBooleans though)

Did you mean variants, like debug/release? Afaik this is a BuildType, and I didn't find a method to disable these build types: https://developer.android.com/reference/tools/gradle-api/8.0/com/android/build/api/dsl/BuildType

Or do you mean a product flavor?

We don't use any special AGP callback/API. Instead, we use `configureEach`. This Gradle API should only run for the actual buildtypes/product flavors. So disabling a product flavor using https://developer.android.com/build/build-variants#filter-variants should work, will add a test later.
But yes, this will still create a disabled task `licenseeFreeGoogle` and the real enabled task `licenseeAndroidFreeGoogle`.

I don't think, the new callbacks would help, due to the AGP + KGP mess with two plugins using `withPlugin`. Additionally, AGP 8 requires Gradle 8, but this isn't fully compatible with KGP 1.8.20...

**Comment by hfhbd** [comment]
Alternatively, we could keep `afterEvaluate` for Android, but remove it for plain projects, like `java`, `kotlin.jvm`, and `kotlin.js`, I think.

**Comment by hfhbd** [comment]
What exactly is your problem with the (annoying) additional created tasks? The visibility in AS/Idea? If so, we could "hide" them by changing the group from `verification` to `other`. Not ideal too, but at least it is less confusing and not visible in `verification` anymore.

**Comment by hfhbd** [comment]
I switched the implementation: Plain/normal plugins (java, kotlin.jvm, kotlin.js, kotlin.mpp) are configured directly (and mpp without Android support).
The Android plugin still uses `afterEvaluate`. This is needed to don't register unused tasks which need to be skipped but are still visible.

I think this is the best implementation (ever).

File: src/main/kotlin/app/cash/licensee/plugin.kt [hunk 1]
```diff
@@ -49,76 +51,95 @@ class LicenseePlugin : Plugin<Project> {
...
+}
+
+private fun handleAndroid(
+  project: Project,
+  appliedAndroidPlugin: AndroidPlugin,
+) {
+  project.afterEvaluate {
+    val withMPP = project.pluginManager.hasPlugin("org.jetbrains.kotlin.multiplatform")
+    if (withMPP) {
+      val rootTask = configureMultiPlatformPlugin(project)
```
Comment (hfhbd) [inline src/main/kotlin/app/cash/licensee/plugin.kt]: Calling `configureMultiPlatformPlugin` again is safe, because the call has no side effects.

**Comment by hfhbd** [comment]
The new `onVariants` api is nice, but requires no `afterEvaluate`.
With this requirement, the only option to support AGP + MPP I think, is prefixing the android tasks and output folders with `android`. This option does not register unnecessary tasks, but would be a breaking change.

File: src/test/kotlin/app/cash/licensee/LicenseePluginFixtureTest.kt [hunk 2]
```diff
@@ -194,7 +194,7 @@ class LicenseePluginFixtureTest {
     val fixtureDir = File(fixturesDir, fixtureName)
     val result = createRunner(fixtureDir).buildAndFail()
     assertThat(result.output).contains(
-      "'app.cash.licensee' requires compatible language/platform plugin to be applied (project :some:thing)",
+      "Task 'licensee' not found in root project 'plugin-missing-on-subproject-fails' and its subprojects.",
```
Comment (JakeWharton) [inline src/test/kotlin/app/cash/licensee/LicenseePluginFixtureTest.kt]: I really don't like this failure mode. Can we keep `afterEvaluate` for detecting this case?

Comment (hfhbd) [reply src/test/kotlin/app/cash/licensee/LicenseePluginFixtureTest.kt]: Sure, I added this check back.

File: src/test/fixtures/plugin-missing-on-root-fails/build.gradle [hunk 3]
```diff
@@ -1,3 +1,4 @@
 plugins {
+  id("base")
```
Comment (JakeWharton) [inline src/test/fixtures/plugin-missing-on-root-fails/build.gradle]: Why is this needed now?

Comment (hfhbd) [reply src/test/fixtures/plugin-missing-on-root-fails/build.gradle]: It's not. Left over from the now reverted failure mode.
